src/dqrng.cpp

```diff
@@ -145,6 +146,33 @@ double rexp(double rate = 1.0) {
   return rexp_impl();
 }
 
+//' @rdname dqrng-functions
+//' @export
+// [[Rcpp::export(rng = false)]]
+Rcpp::IntegerVector dqrrademacher(size_t n) {
+  size_t n_ints = ceil(n / 64.0);
+  std::vector<uint64_t> rand_ints(n_ints);
+  std::generate(rand_ints.begin(), rand_ints.end(), ruint64t_impl);
+
+  Rcpp::IntegerVector res(n);
+  int k = 0;
+  for (int i = 0; i < n_ints - 1; ++i) {
+    uint64_t curr = rand_ints[i];
+    
+    for (int j = 63; j >= 0; j--) {
+      res[k] = ((curr >> j) & 1) * 2 - 1;
+      k++;
+    }
+  }
+
+  uint64_t curr = rand_ints[n_ints - 1];
+  for (int j = 63; k < n; ++k, --j) {
+    res[k] = ((curr >> j) & 1) * 2 - 1;    
+  }
+
+  return res;
+}
```

Comment:
I think one can boil this down to

```
//' @rdname dqrng-functions
//' @export
// [[Rcpp::export(rng = false)]]
Rcpp::IntegerVector dqrrademacher(size_t n) {
  Rcpp::IntegerVector res(n);
  size_t k = 0;
  while (k < n) {
    uint64_t bits = (*rng)();

    for (int j = 0; j < 64 && k < n; ++k, ++j) {
      res[k] = ((bits >> j) & 1) * 2 - 1;
    }
  }
  return res;
}
```

* There is no need to pre-compute the (number of) random draws. We just generate them as needed.
* Any reason why you had the `j` loop decreasing? I prefer increasing loops unless there is a good reason otherwise.

what do you think?

Reply:
Very clean code! I didn't think to do that. It seems from benchmarking, that the while loop is 1/2 as fast. I think it has to do with compiler optimizations b/c when I use `devtools::load_all()` which doesn't optimize, they run at the same speed. I assume it's from having to check the statement `n` times

```cpp

//' @rdname dqrng-functions
//' @export
// [[Rcpp::export(rng = false)]]
Rcpp::IntegerVector dqrrademacher(size_t n) {
  Rcpp::IntegerVector res = Rcpp::no_init(n);
  size_t k = 0;
  for (size_t i = 0; i < ceil(n / 64.0) - 1; ++i) {
    uint64_t curr = (*rng)();
    
    for (int j = 0; j <= 63; ++j) {
      res[k] = ((curr >> j) & 1) * 2 - 1;
      k++;
    }
  }

  uint64_t curr = (*rng)();
  for (int j = 0; k < n; ++k, ++j) {
    res[k] = ((curr >> j) & 1) * 2 - 1;    
  }

  return res;
}

//' @rdname dqrng-functions
//' @export
// [[Rcpp::export(rng = false)]]
Rcpp::IntegerVector dqrrademacher_v2(size_t n) {
  Rcpp::IntegerVector res = Rcpp::no_init(n);
  size_t k = 0, j;
  uint64_t bits;
  while (k < n) {
    j = k % 64;
    if (j == 0) bits = (*rng)();
    res[k] = ((bits >> j) & 1) * 2 - 1;
    ++k;
  }
  return res;
}
```

``` r
library(dqrng)
library(rademacher)
n = 1e7
bench::mark(
  dqrrademacher = dqrrademacher(n),
  dqrrademacher_v2 = dqrrademacher_v2(n),
  rademacher = rademacher::sample_rademacher(n),
  runif = (stats::runif(n) > 0.5) * 2 - 1,
  samp = sample(x = c(1, -1), size = n, replace = TRUE),
  dqsamp = dqsample(x = c(1,-1), size = n, replace = TRUE),
  check = FALSE,
  iterations = 100
)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 6 × 6
#>   expression            min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>       <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 dqrrademacher      3.74ms   5.23ms    198.      38.1MB    96.8 
#> 2 dqrrademacher_v2    7.2ms   8.48ms    108.      38.1MB    32.3 
#> 3 rademacher         6.44ms   6.94ms    130.      39.4MB    20.8 
#> 4 runif             87.43ms  90.02ms     11.0    190.7MB    11.0 
#> 5 samp             104.41ms 108.54ms      9.15   114.4MB     7.23
#> 6 dqsamp            43.67ms   47.7ms     20.7    114.5MB    16.6
```

<sup>Created on 2023-07-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
I did change the function to use `(*rng)()` as needed, though! Good idea

Reply:
Interesting. In that case it makes sense to handle it like this.